go/parquet/internal/utils/clib.go

```diff
@@ -0,0 +1,30 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !noasm && !appengine && (amd64 || arm64)
+// +build !noasm
+// +build !appengine
+// +build amd64 arm64
+
+package utils
+
+import "unsafe"
+
+//go:noescape
+func _ClibMemcpy(dst, src unsafe.Pointer, n uint) unsafe.Pointer
+
+//go:noescape
+func _ClibMemset(dst unsafe.Pointer, c int, n uint) unsafe.Pointer
```

Comment:
I'm curious, what is this and what is it good for? Doesn't Go provide memcpy/memset replacements?

Reply:
this exists *purely* to test the assembly versions I provided. 

The issue is that the C implementation I'm using to generate the assembly makes calls to memset and memcpy, but C's calling conventions are different than Go's. So in the assembly when it calls memset/memcpy I replaced those CALL instructions to call these replacement versions (because I can't call out to libc from inside of Go like that, no guarantee that the lib is loaded). Which are essentially just ported versions of the C memcpy/memset to ensure that the Calling conventions and argument handling are the same. 

The underscore prefix there ensures that the methods are not exported from the package, so I'm only using these for `clib_test.go` in order to sanity check the assembly methods.

Reply:
Yuck, is there a way to avoid the calls to memset and memcpy?

Reply:
Even the original C++ implementations (bpacking_simd128_generated.h) use memset and memcpy. There might be ways to avoid those calls but they'd likely be much less efficient. We also already use amd64 versions of these anyways, so it's good to have arm64 implementations.

Reply:
By the looks of it, the calls to memcpy/memset could easily be replaced by the equivalent `for` loop. Given the size is small and constant, I don't think calling a generic version of memcpy/memset would be a significant win.

Reply:
I'll give it a try

Reply:
Latest update has removed the `memset`/`memcpy` and replaced them with simple fixed size loops. My local testing showed all the tests passing, so assuming the CI is happy with things then we should be good here if you're happy with the results @pitrou. Let me know please when you get a chance.